### PR TITLE
feat: Glue Catalog - table operations (3/3)

### LINKED ISSUE #249: Implement Glue Catalog
Implement glue catalog.

### PR DESCRIPTION
**Pull request by marvinlanhenke**
### Which issue does this PR close?
Closes #249 (Task 3/3)

### Rationale for this change
Add support for Glue Catalog, to reach feature parity with other implementations.

### What changes are included in this PR?
- add table operations
- except `fn update_table()`

### Are these changes tested?
Yes. Unit and integration tests are included.

### CODE REVIEW

**Comment by marvinlanhenke** [comment]
@liurenjie1024 PTAL

File: crates/catalog/glue/src/catalog.rs [hunk 1]
```diff
@@ -310,31 +327,282 @@ impl Catalog for GlueCatalog {
...
+        match glue_table_output.table() {
+            None => Err(Error::new(
+                ErrorKind::Unexpected,
+                format!(
+                    "'Table' object for database: {} and table: {} does not exist",
+                    db_name, table_name
+                ),
+            )),
+            Some(table) => {
+                let metadata_location = get_metadata_location(&table.parameters)?;
```
Comment (liurenjie1024) [inline crates/catalog/glue/src/catalog.rs:432]: Python implementation have some check about glue table: https://github.com/apache/iceberg-python/blob/07442cc00125ac00a91f717d9832e5479f4ff6dd/pyiceberg/catalog/glue.py#L299

But I think this is not a blocker.

File: crates/catalog/glue/src/catalog.rs [hunk 2]
```diff
@@ -310,31 +327,282 @@ impl Catalog for GlueCatalog {
...
+                    .create_table()
+                    .database_name(&dest_db_name)
+                    .table_input(rename_table_input);
+                let builder = with_catalog_id!(builder, self.config);
+
+                builder.send().await.map_err(from_aws_sdk_error)?;
+
+                let drop_src_table_result = self.drop_table(src).await;
+
+                match drop_src_table_result {
```
Comment (liurenjie1024) [inline crates/catalog/glue/src/catalog.rs:570]: This looks great, thanks!

File: crates/catalog/glue/tests/glue_catalog_test.rs [hunk 3]
```diff
@@ -91,6 +103,166 @@ async fn set_test_namespace(fixture: &TestFixture, namespace: &NamespaceIdent) -
...
+#[tokio::test]
+async fn test_table_exists() -> Result<()> {
+    let fixture = set_test_fixture("test_table_exists").await;
+    let creation = set_table_creation("s3a://warehouse/hive", "my_table")?;
+    let namespace = Namespace::new(NamespaceIdent::new("my_database".into()));
+
+    fixture
+        .glue_catalog
+        .create_namespace(namespace.name(), HashMap::new())
+        .await?;
```
Comment (liurenjie1024) [inline crates/catalog/glue/tests/glue_catalog_test.rs:163]: How about adding a check here to ensure that `table_exists` returns false?

File: crates/catalog/glue/tests/glue_catalog_test.rs [hunk 4]
```diff
@@ -91,6 +103,166 @@ async fn set_test_namespace(fixture: &TestFixture, namespace: &NamespaceIdent) -
...
+
+    let dest = TableIdent::new(namespace.name().clone(), "my_table_rename".to_string());
+
+    fixture
+        .glue_catalog
+        .rename_table(table.identifier(), &dest)
+        .await?;
+
+    let table = fixture.glue_catalog.load_table(&dest).await?;
+    assert_eq!(table.identifier(), &dest);
```
Comment (liurenjie1024) [inline crates/catalog/glue/tests/glue_catalog_test.rs:149]: How about adding a check here to load old table name and verify that it doesn't exist?

**Review by liurenjie1024 — COMMENTED** [review]
Thanks for @marvinlanhenke for this great pr, it looks great! Sorry for late reply, I have been busy lately.

File: crates/catalog/glue/src/utils.rs [hunk 5]
```diff
@@ -151,6 +205,65 @@ pub(crate) fn validate_namespace(namespace: &NamespaceIdent) -> Result<String> {
...
 
+/// Get default table location from `Namespace` properties
+pub(crate) fn get_default_table_location(
+    namespace: &Namespace,
+    table_name: impl AsRef<str>,
+    warehouse: impl AsRef<str>,
+) -> String {
+    let properties = namespace.properties();
+
+    let location = match properties.get(LOCATION) {
```
[the remark this answers is not in the document]
Comment (marvinlanhenke) [reply crates/catalog/glue/src/utils.rs:216]: Yes, you're right. 

It seems that I missed `{database}.db` which would, in our case, equate to the `db_name` (NamespaceIdent(Vec<String>)).

Should be easy to fix. Thanks for spotting this.

File: crates/catalog/glue/src/catalog.rs [hunk 6]
```diff
@@ -310,31 +327,282 @@ impl Catalog for GlueCatalog {
...
+            .name(table_name);
+        let builder = with_catalog_id!(builder, self.config);
+
+        let glue_table_output = builder.send().await.map_err(from_aws_sdk_error)?;
+
+        match glue_table_output.table() {
+            None => Err(Error::new(
+                ErrorKind::Unexpected,
+                format!(
+                    "'Table' object for database: {} and table: {} does not exist",
```
[the remark this answers is not in the document]
Comment (marvinlanhenke) [reply crates/catalog/glue/src/catalog.rs:427]: Probably wanted to remove both single quotes here?

File: crates/catalog/glue/src/catalog.rs (hunk 1 above)
Comment (marvinlanhenke) [reply crates/catalog/glue/src/catalog.rs:432]: Yes, I kind of 'ignored' this on purpose - same check is missing for the hive catalog as well; 
I guess we can handle those in separate PRs.

**Comment by marvinlanhenke** [comment]
> Thanks for @marvinlanhenke for this great pr, it looks great! Sorry for late reply, I have been busy lately.

@liurenjie1024 
No worries, and thanks for the review.

I'll most likely fix those remaining suggestions tomorrow.

**Comment by marvinlanhenke** [comment]
> For the 'missing table checks' - I'll file an Issue later, so we can track this and implement in separate PRs.

#337

**Review by liurenjie1024 — APPROVED** [review]
LGTM, thanks @marvinlanhenke for this pr!

**Comment by liurenjie1024** [comment]
cc @Fokko @Xuanwo @sdd PTAL

**Comment by liurenjie1024** [comment]
Thanks @marvinlanhenke for this pr!

**Pull request merged**